pinot-broker/src/main/java/org/apache/pinot/broker/routing/segmentpruner/PartitionSegmentPruner.java

```diff
@@ -163,48 +181,44 @@ public synchronized void refreshSegment(String segment) {
       if (filterExpression == null) {
         return segments;
       }
-      Set<String> selectedSegments = new HashSet<>();
-      for (String segment : segments) {
-        PartitionInfo partitionInfo = _partitionInfoMap.get(segment);
-        if (partitionInfo == null || partitionInfo == INVALID_PARTITION_INFO || isPartitionMatch(filterExpression,
-            partitionInfo)) {
-          selectedSegments.add(segment);
-        }
-      }
-      return selectedSegments;
+      return pruneSegments((partitionInfo, cachedPartitionFunction) -> isPartitionMatch(filterExpression,
```

Comment:
(Critical) This algorithm does not prune based on the input segments, and will ignore the results for other pruners. Also, we need to keep the segments with `null` or `INVALID_PARTITION_INFO` because these segments might not be partitioned

Reply:
Yes you are right. Found out in the failed unit tests. Changed and updated now, but needs another round of performance testing.
Will ping you folks for review after another round of testing is done.